source/brailleDisplayDrivers/nattiqbraille.py

```diff
@@ -0,0 +1,147 @@
+ROUTE_RESP = 1
+UP_KEY_RESP = 2
+DOWN_KEY_RESP = 3
+RIGHT_KEY_RESP = 4
+LEFT_KEY_RESP = 5
+# Keys pressed id
+UP_KEY_PRESS = 1
+DOWN_KEY_PRESS = 2
+RIGHT_KEY_PRESS = 3
+LEFT_KEY_PRESS = 4
+
+
+class BrailleDisplayDriver(braille.BrailleDisplayDriver):
+	name = "nattiqbraille"
+	# Translators: Names of braille displays
+	description = _("Nattiq nBraille")
+	isThreadSafe = True
+
+	@classmethod
+	def getManualPorts(cls):
+		return braille.getSerialPorts()
+
+	def __init__(self, port="auto"):
+		super(BrailleDisplayDriver, self).__init__()
+		self._serial = None
+		for portType, portId, port, portInfo in self._getTryPorts(port):
+			log.debug("Checking port %s for a Nattiq nBraille", port)
+			try:
+				self._serial = hwIo.Serial(
+					port, baudrate=BAUD_RATE, timeout=self.timeout, writeTimeout=self.timeout,
+					parity=serial.PARITY_NONE, onReceive=self._onReceive
+				)
+			except EnvironmentError:
+				log.debugWarning("", exc_info=True)
+				continue
+			# Check for cell information
+			if self._describe():
+				log.debug("Nattiq nBraille found on %s with %d cells", port, self.numCells)
+				break
+			else:
+				self._serial.close()
+		else:
+			raise RuntimeError("Can't find a Nattiq nBraille device (port = %s)" % port)
+
+	def terminate(self):
+		try:
+			super(BrailleDisplayDriver, self).terminate()
+		finally:
+			self._serial.write(RESET_TAG)
+			self._serial.close()
+			self._serial = None
+
+	def _describe(self):
+		self.numCells = 0
+		log.debug("Writing reset tag")
+		self._serial.write(RESET_TAG)
+		self._serial.waitForRead(self.timeout * 10)
+		log.debug("Writing init tag")
+		self._serial.write(INIT_TAG)
+		self._serial.waitForRead(self.timeout * 10)
+		# If a valid response was received, _onReceive will have set numCells.
+		if self.numCells:
+			return True
+		log.debug("Not a Nattiq nBraille")
+		return False
+
+	def _onReceive(self, command):
+		if int(command) == INIT_RESP:
+			CELLS_NUM = self._serial.read(2)
+			self.numCells = int(CELLS_NUM)
+		elif int(command) == ROUTE_RESP:
+			ROUTE_KEY = self._serial.read(2)
+			inputCore.manager.executeGesture(RoutingInputGesture(int(ROUTE_KEY)))
+		elif int(command) == UP_KEY_RESP:
+			inputCore.manager.executeGesture(InputGestureKeys(UP_KEY_PRESS))
+			self._serial.waitForRead(self.timeout * 10)
+			log.debug("Up Key Pressed")
+		elif int(command) == DOWN_KEY_RESP:
+			inputCore.manager.executeGesture(InputGestureKeys(DOWN_KEY_PRESS))
+			self._serial.waitForRead(self.timeout * 10)
+			log.debug("Down Key Pressed")
+		elif int(command) == RIGHT_KEY_RESP:
+			inputCore.manager.executeGesture(InputGestureKeys(RIGHT_KEY_PRESS))
+			self._serial.waitForRead(self.timeout * 10)
+			log.debug("Right Key Pressed")
+		elif int(command) == LEFT_KEY_RESP:
+			inputCore.manager.executeGesture(InputGestureKeys(LEFT_KEY_PRESS))
+			self._serial.waitForRead(self.timeout * 10)
+			log.debug("Left Key Pressed")
+
+	def display(self, cells):
+		cells = b"-".join(str(cell) for cell in cells)

```

Comment:
Are you really sure that this works? You're mixing str with bytes here. Please see my earlier suggestion.

Reply:
I edited out to your suggestion, I will test the build again,
from what I understood, that it does not matter if NVDA sent the cells data as bytes or as a string, because whether they were this or that, they will be converted to integers for the Braille display. so mostly it should not matter, but I need to test again with the unit I have and see how it performs

Reply:
Nope, with converting the cells data to bytes, it does not works. It gets detected fine, and input is OK, but no output. So it seems it requires the cells data (0-255) to be in String format.
Do you suggest to keep this 
b"-".join(str(cell) for cell in cells)
or to keep the dash in string?
"-".join(str(cell) for cell in cells)

Both works fine after testings, because the dash "-" is only a separator and it gets ignored. As the important part is the cells data to display them for each cell.
Thanks.

Reply:
How important are these dashes? Are they prescribed by the protocol?

Reply:
> How important are these dashes? Are they prescribed by the protocol?

Yes, they are separating the data for the cells. The unit receives the whole line one time instead of feeding it one by one, so it needs to have a separator of the cell data.
Then after the unit receive the line data, it converts each cell string to its byte value (8-bit cells for 8-dots) which means the (0-255) gets into each cell by the order it gets from NVDA.